docs: adds article for OAuth2 and OIDC disovery endpoints (#2324)
ory/docs: docs/oauth2-oidc/wellknown-endpoint-discovery.mdx

---
id: wellknown-endpoint-discovery
title: OAuth2.0 & OpenID Connect well-known endpoint discovery

> sidebar_label: OAuth2.0 & OpenID Connect well-known endpoint discovery

Author: Yes, I was using the term TOC/sidebar nav interchangeably

---

There are two types of well-known discovery endpoints. The OAuth2.0 and the OpenID Connect (ODIC) discovery endpoints. Both

> endpoints return nearly identical information as OIDC is built on top of OAuth 2.0 (OIDC is a superset of OAuth 2.0).

Reviewer: ```suggestion
endpoints return nearly identical information.
```

imo we dont need to explain OIDC/OAuth2 here, I would assume the reader already know this - alternatively link to the document that explains this in detail?

The discovery endpoints provide a standardized way for clients to automatically discover your server's configuration and
capabilities. They eliminate manual configuration, enable dynamic multi-tenant support, and make OAuth 2.0 and ODIC integrations
more resilient to changes. Discovery endpoints allow clients to adapt automatically to configuration changes without code updates.

## OAuth 2.0's discovery endpoint

> Use OAuth 2.0's `oauth-authorization-server` when you're implementing pure OAuth 2.0 (authorization only). Example use cases are

Reviewer: ```suggestion
## OAuth 2.0 discovery endpoint

Use the OAuth 2.0 `oauth-authorization-server` when you're implementing pure OAuth 2.0 (authorization only). Example use cases are
```

see below for context

when you need: - API access tokens - Machine-to-machine authentication - Resource server authorization - Client credentials flow -

> No user identity information needed

Reviewer: ```suggestion
when you need: 
- API access tokens 
- Machine-to-machine authentication 
- Resource server authorization 
- Client credentials flow 
- No user identity information needed
```

The list here is broken - please always check how the page renders either locally before you open the PR or in the preview when it is pushed and in draft stage

## OIDC's discovery endpoint

Use OIDC's `openid-configuration` when you're implementing OpenID Connect (authentication + authorization). Example use cases are
when you need: - User login/authentication - User profile information - ID tokens - Single sign-on (SSO) - Other sign in flows

Click the appropriate tab below to learn more about the OAuth2.0 or OIDC discovery endpoint.

<Tabs>
  <TabItem value="OAuth 2" label="OAuth 2.0 discovery endpoint" default>

## What is the OAuth 2.0 discovery endpoint?

The discovery endpoint returns a JSON document containing metadata about your OAuth 2.0 authorization server, including:

- Available endpoints (authorization, token, userinfo, etc.)
- Supported grant types and response types
- Supported authentication methods
- Available scopes
- Security capabilities (PKCE, token revocation, etc.)

## Implementation with Ory

### Ory Network

Ory Network automatically provides the discovery endpoint for your project:

```bash
# Replace with your project slug
curl https://your-project.projects.oryapis.com/.well-known/oauth-authorization-server
```

> ### Ory Hydra (Self-hosted)

Author: Until we've a better system to distinguish the products, yes I prefer to keep it.

1. Client performs authentication using the discovered endpoints

### Example discovery requests

```bash
curl https://your-project.projects.oryapis.com/.well-known/oauth-authorization-server | jq
```

### Example response

```json
{
  "issuer": "https://your-project.projects.oryapis.com",
  "authorization_endpoint": "https://your-project.projects.oryapis.com/oauth2/auth",
  "token_endpoint": "https://your-project.projects.oryapis.com/oauth2/token",
  "jwks_uri": "https://your-project.projects.oryapis.com/.well-known/jwks.json",
  "userinfo_endpoint": "https://your-project.projects.oryapis.com/userinfo",
  "revocation_endpoint": "https://your-project.projects.oryapis.com/oauth2/revoke",
  "introspection_endpoint": "https://your-project.projects.oryapis.com/oauth2/introspect",
  "grant_types_supported": ["authorization_code", "refresh_token", "client_credentials"],
  "response_types_supported": ["code", "token", "id_token"],
  "scopes_supported": ["openid", "offline_access", "profile", "email"],
  "token_endpoint_auth_methods_supported": ["client_secret_post", "client_secret_basic", "private_key_jwt", "none"],
  "code_challenge_methods_supported": ["S256", "plain"]
}

> ```

Reviewer: To keep it consistent with the rest of the documentation, can you please use the placeholder like described [here](https://github.com/ory/docs/?tab=readme-ov-file#placeholders-and-dummy-data)

>When referring to a project API URL or SDK URL, always use https://$PROJECT_SLUG.projects.oryapis.com.

> ```

Author: did the replacement!

- `revocation_endpoint`: URL to revoke tokens
- `introspection_endpoint`: URL to validate tokens
- `registration_endpoint`: URL for dynamic client registration
- `scopes_supported`: List of supported OAuth 2.0 scopes
- `response_types_supported`: Supported response types
- `grant_types_supported`: Supported grant types
- `token_endpoint_auth_methods_supported`: Client authentication methods
- `code_challenge_methods_supported`: PKCE methods (S256, plain)

## Best practices

### Cache discovery results

Discovery documents change infrequently. You can cache them to reduce latency:

```js
const CACHE_KEY = "oauth_discovery"
const CACHE_DURATION = 24 * 60 * 60 * 1000 // 24 hours

async function getDiscoveryDocument(issuerUrl) {
  const cached = localStorage.getItem(CACHE_KEY)
  if (cached) {
    const { data, timestamp } = JSON.parse(cached)
    if (Date.now() - timestamp < CACHE_DURATION) {
      return data
    }
  }

  const discovery = await fetch(`${issuerUrl}/.well-known/oauth-authorization-server`).then((r) => r.json())

  localStorage.setItem(
    CACHE_KEY,
    JSON.stringify({
      data: discovery,
      timestamp: Date.now(),
    }),
  )

  return discovery
}
```

> ### Handle errors gracefully

Author: Same comment about toc above. Will fix in restructure.

```js
async function fetchDiscovery(issuerUrl) {
  try {
    const response = await fetch(`${issuerUrl}/.well-known/oauth-authorization-server`)

    if (!response.ok) {
      throw new Error(`Discovery failed: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error("Failed to fetch discovery document:", error)
    // Fall back to manual configuration or show error to user
    throw error
  }
}
```

### Validate required fields

```js
function validateDiscovery(config) {
  const required = ["issuer", "authorization_endpoint", "token_endpoint", "jwks_uri"]

  for (const field of required) {
    if (!config[field]) {
      throw new Error(`Missing required field: ${field}`)
    }
  }

  return true
}
```

## Troubleshooting

> ### Discovery endpoint returns 404

Reviewer: ```suggestion
##### Discovery endpoint returns 404
```

see above ToC too crowded

  "jwks_uri": "https://your-project.projects.oryapis.com/.well-known/jwks.json",
  "registration_endpoint": "https://your-project.projects.oryapis.com/oauth2/register",
  "revocation_endpoint": "https://your-project.projects.oryapis.com/oauth2/revoke",
  "introspection_endpoint": "https://your-project.projects.oryapis.com/oauth2/introspect",
  "end_session_endpoint": "https://your-project.projects.oryapis.com/oauth2/sessions/logout",
  "response_types_supported": ["code", "code id_token", "id_token", "token id_token", "token", "token id_token code"],
  "subject_types_supported": ["public", "pairwise"],
  "id_token_signing_alg_values_supported": ["RS256", "ES256"],
  "scopes_supported": ["openid", "offline_access", "profile", "email"],
  "token_endpoint_auth_methods_supported": ["client_secret_post", "client_secret_basic", "private_key_jwt", "none"],
  "claims_supported": ["sub", "email", "email_verified", "name", "given_name", "family_name", "picture"],
  "code_challenge_methods_supported": ["S256", "plain"],
  "grant_types_supported": ["authorization_code", "implicit", "refresh_token", "client_credentials"]
}

> ```

Reviewer: I get a somewhat different response when I try this on a fresh Ory Network project:

```json
{
  "issuer": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com",
  "authorization_endpoint": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/oauth2/auth",
  "device_authorization_endpoint": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/oauth2/device/auth",
  "token_endpoint": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/oauth2/token",
  "jwks_uri": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/.well-known/jwks.json",
  "subject_types_supported": [
    "public"
  ],
  "response_types_supported": [
    "code",
    "code id_token",
    "id_token",
    "token id_token",
    "token",
    "token id_token code"
  ],
  "claims_supported": [
    "sub"
  ],
  "grant_types_supported": [
    "authorization_code",
    "implicit",
    "client_credentials",
    "refresh_token",
    "urn:ietf:params:oauth:grant-type:device_code"
  ],
  "response_modes_supported": [
    "query",
    "fragment",
    "form_post"
  ],
  "userinfo_endpoint": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/userinfo",
  "scopes_supported": [
    "offline_access",
    "offline",
    "openid"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_post",
    "client_secret_basic",
    "private_key_jwt",
    "none"
  ],
  "userinfo_signing_alg_values_supported": [
    "none",
    "RS256"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "id_token_signed_response_alg": [
    "RS256"
  ],
  "userinfo_signed_response_alg": [
    "RS256"
  ],
  "request_parameter_supported": true,
  "request_uri_parameter_supported": true,
  "require_request_uri_registration": true,
  "claims_parameter_supported": false,
  "revocation_endpoint": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/oauth2/revoke",
  "backchannel_logout_supported": true,
  "backchannel_logout_session_supported": true,
  "frontchannel_logout_supported": true,
  "frontchannel_logout_session_supported": true,
  "end_session_endpoint": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/oauth2/sessions/logout",
  "request_object_signing_alg_values_supported": [
    "none",
    "RS256",
    "ES256"
  ],
  "code_challenge_methods_supported": [
    "plain",
    "S256"
  ],
  "credentials_endpoint_draft_00": "https://naughty-bouman-acdevc3rq2.projects.oryapis.com/credentials",
  "credentials_supported_draft_00": [
    {
      "format": "jwt_vc_json",
      "types": [
        "VerifiableCredential",
        "UserInfoCredential"
      ],
      "cryptographic_binding_methods_supported": [
        "jwk"
      ],
      "cryptographic_suites_supported": [
        "PS256",
        "RS256",
        "ES256",
        "PS384",
        "RS384",
        "ES384",
        "PS512",
        "RS512",
        "ES512",
        "EdDSA"
      ]
    }
  ]
}
```

- `token_endpoint_auth_methods_supported`: Client authentication methods
- `code_challenge_methods_supported`: PKCE methods (S256, plain)
- `end_session_endpoint`: URL for logout/end session
- `revocation_endpoint`: URL to revoke tokens
- `introspection_endpoint`: URL to validate tokens

## Best practices

### Cache discovery results

Discovery documents change infrequently. Cache them to improve performance:

```javascript
class CachedOIDCClient {
  static CACHE_KEY = "oidc_discovery"
  static CACHE_DURATION = 24 * 60 * 60 * 1000 // 24 hours

  static async getDiscovery(issuerUrl) {
    const cached = localStorage.getItem(this.CACHE_KEY)

    if (cached) {
      const { data, timestamp, issuer } = JSON.parse(cached)
      if (issuer === issuerUrl && Date.now() - timestamp < this.CACHE_DURATION) {
        return data
      }
    }

    const discoveryUrl = `${issuerUrl}/.well-known/openid-configuration`
    const discovery = await fetch(discoveryUrl).then((r) => r.json())

    localStorage.setItem(
      this.CACHE_KEY,
      JSON.stringify({
        data: discovery,
        timestamp: Date.now(),
        issuer: issuerUrl,
      }),
    )

    return discovery
  }
}
```

> ### Validate discovery response

Reviewer: ```suggestion
##### Validate discovery response
```
see above ToC too crowded

Author: Same comment about toc above. Will fix in restructure.

    // Fall back to manual configuration or show error
    throw new Error(`Failed to discover OIDC configuration: ${error.message}`)
  }
}
```

### Verify issuer match

Always verify the issuer in the discovery document matches your expected issuer:

```javascript
const discovery = await fetch(discoveryUrl).then((r) => r.json())

if (discovery.issuer !== expectedIssuer) {
  throw new Error("Issuer mismatch - possible security issue")
}
```

## Troubleshooting

### Discovery endpoint returns 404

Verify the URL format:

```text
https://your-auth-server.com/.well-known/openid-configuration
```

For Ory Network, ensure you're using your project's full URL:

```text
https://your-project.projects.oryapis.com/.well-known/openid-configuration
```

### Issuer mismatch error

Ensure the issuer in the discovery document matches your expected issuer URL exactly (including trailing slashes):

```javascript
// Discovery returns
{ "issuer": "https://auth.example.com" }

// Your code expects
const expectedIssuer = "https://auth.example.com"; // ✓ Match

// NOT
const expectedIssuer = "https://auth.example.com/"; // ✗ Trailing slash mismatch
```

> ### CORS issues

Reviewer: ```suggestion
##### CORS issues
```
see above ToC too crowded

If calling discovery from a browser, ensure CORS is configured. Ory Network handles this automatically.

For self-hosted Ory Hydra:

```yaml
# hydra.yml
serve:
  public:
    cors:
      enabled: true
      allowed_origins:
        - https://your-app.com
      allowed_methods:
        - GET
        - POST
        - OPTIONS
```

> ### Missing required scopes

Reviewer: ```suggestion
##### Missing required scopes
```
see above ToC too crowded
